File: lib/startup/customizeApplication.js
const fs = require('fs')
const config = require('config')
const replace = require('replace')
const utils = require('../utils')

const customizeApplication = () => {
  if (config.get('application.name')) {
    customizeTitle()
  }
  if (config.get('application.logo')) {
    customizeLogo()
  }
  if (config.get('application.favicon')) {
    customizeFavicon()
  }
  if (config.get('application.theme')) {
    customizeTheme()
  }
  if (config.get('application.cookieConsent')) {
    customizeCookieConsentBanner()
  }
  if (config.get('application.promotion')) {
    customizePromotionVideo()
    customizePromotionSubtitles()
  }
  if (config.get('hackingInstructor')) {
    customizeHackingInstructorAvatar()
  }
}

const customizeLogo = () => {
  retrieveCustomFile('application.logo', 'frontend/dist/frontend/assets/public/images')
}

const customizeFavicon = () => {
  const favicon = retrieveCustomFile('application.favicon', 'frontend/dist/frontend/assets/public')
  replace({
    regex: /type="image\/x-icon" href="assets\/public\/.*"/,
    replacement: `type="image/x-icon" href="assets/public/${favicon}"`,
    paths: ['frontend/dist/frontend/index.html'],
    recursive: false,
    silent: true
  })
}

const customizePromotionVideo = () => {
  retrieveCustomFile('application.promotion.video', 'frontend/dist/frontend/assets/public/videos')
}

const customizePromotionSubtitles = () => {
  retrieveCustomFile('application.promotion.subtitles', 'frontend/dist/frontend/assets/public/videos')
}

const retrieveCustomFile = (sourceProperty, destinationFolder) => {
  let file = config.get(sourceProperty)
  if (utils.startsWith(file, 'http')) {
    const filePath = file
    file = utils.extractFilename(file)
    utils.downloadToFile(filePath, destinationFolder + '/' + file)
  }
  return file
}

const customizeTitle = () => {
  const title = `<title>${config.get('application.name')}</title>`
  replace({
    regex: /<title>.*<\/title>/,
    replacement: title,
    paths: ['frontend/dist/frontend/index.html'],
    recursive: false,
    silent: true
  })
}

const customizeTheme = () => {
  const bodyClass = '"mat-app-background ' + config.get('application.theme') + '-theme"'
  replace({
    regex: /"mat-app-background .*-theme"/,
    replacement: bodyClass,
    paths: ['frontend/dist/frontend/index.html'],
    recursive: false,
    silent: true
  })
}

const customizeCookieConsentBanner = () => {
  const popupProperty = '"popup": { "background": "' + config.get('application.cookieConsent.backgroundColor') + '", "text": "' + config.get('application.cookieConsent.textColor') + '" }'
  replace({
    regex: /"popup": { "background": ".*", "text": ".*" }/,
    replacement: popupProperty,
    paths: ['frontend/dist/frontend/index.html'],
    recursive: false,
    silent: true
  })
  const buttonProperty = '"button": { "background": "' + config.get('application.cookieConsent.buttonColor') + '", "text": "' + config.get('application.cookieConsent.buttonTextColor') + '" }'
  replace({
    regex: /"button": { "background": ".*", "text": ".*" }/,
    replacement: buttonProperty,
    paths: ['frontend/dist/frontend/index.html'],
    recursive: false,
    silent: true
  })
  const contentProperty = '"content": { "message": "' + config.get('application.cookieConsent.message') + '", "dismiss": "' + config.get('application.cookieConsent.dismissText') + '", "link": "' + config.get('application.cookieConsent.linkText') + '", "href": "' + config.get('application.cookieConsent.linkUrl') + '" }'
  replace({
    regex: /"content": { "message": ".*", "dismiss": ".*", "link": ".*", "href": ".*" }/,
    replacement: contentProperty,
    paths: ['frontend/dist/frontend/index.html'],
    recursive: false,
    silent: true
  })
}

const customizeHackingInstructorAvatar = () => {
  let originAvatarImage = 'juicyBot.png'
  if (config.get('hackingInstructor.avatarImage') !== null) {
    originAvatarImage = retrieveCustomFile('hackingInstructor.avatarImage', 'frontend/dist/frontend/assets/public/images')
  }
  fs.copyFileSync('frontend/dist/frontend/assets/public/images/' + originAvatarImage, 'frontend/dist/frontend/assets/public/images/hackingInstructor.png')
}

module.exports = customizeApplication

File: lib/startup/customize_application.py
import os
import requests
import json
import logging
import re

logger = logging.getLogger(__name__)

def customize_application():
    if get_config('application.name'):
        customize_title()
    if get_config('application.logo'):
        customize_logo()
    if get_config('application.favicon'):
        customize_favicon()
    if get_config('application.theme'):
        customize_theme()
    if get_config('application.cookieConsent'):
        customize_cookie_consent_banner()
    if get_config('application.promotion'):
        customize_promotion_video()
        customize_promotion_subtitles()
    if get_config('hackingInstructor'):
        customize_hacking_instructor_avatar()
    if get_config('application.chatBot'):
        customize_chatbot_avatar()

def get_config(key):
    with open('config.json') as f:
        config = json.load(f)
    return config.get(key)

def customize_logo():
    retrieve_custom_file('application.logo', 'frontend/dist/frontend/assets/public/images')

def customize_chatbot_avatar():
    avatar_image = retrieve_custom_file('application.chatBot.avatar', 'frontend/dist/frontend/assets/public/images')
    os.rename('frontend/dist/frontend/assets/public/images/' + avatar_image, 'frontend/dist/frontend/assets/public/images/ChatbotAvatar.png')

def customize_hacking_instructor_avatar():
    avatar_image = retrieve_custom_file('hackingInstructor.avatarImage', 'frontend/dist/frontend/assets/public/images')
    os.rename('frontend/dist/frontend/assets/public/images/' + avatar_image, 'frontend/dist/frontend/assets/public/images/hackingInstructor.png')

def customize_favicon():
    favicon = retrieve_custom_file('application.favicon', 'frontend/dist/frontend/assets/public')
    replace_in_file('frontend/dist/frontend/index.html', r'type="image/x-icon" href="assets/public/.*"', f'type="image/x-icon" href="assets/public/{favicon}"')

def customize_promotion_video():
    retrieve_custom_file('application.promotion.video', 'frontend/dist/frontend/assets/public/videos')

def customize_promotion_subtitles():
    retrieve_custom_file('application.promotion.subtitles', 'frontend/dist/frontend/assets/public/videos')

def retrieve_custom_file(source_property, destination_folder):
    file = get_config(source_property)
    if is_url(file):
        file_path = file
        file = extract_filename(file)
        download_to_file(file_path, os.path.join(destination_folder, file))
    return file

def customize_title():
    title = f'<title>{get_config("application.name")}</title>'
    replace_in_file('frontend/dist/frontend/index.html', r'<title>.*</title>', title)

def customize_theme():
    body_class = f'"mat-app-background {get_config("application.theme")}-theme"'
    replace_in_file('frontend/dist/frontend/index.html', r'"mat-app-background .*-theme"', body_class)

def customize_cookie_consent_banner():
    content_property = f'"content": {{"message": "{get_config("application.cookieConsent.message")}", "dismiss": "{get_config("application.cookieConsent.dismissText")}", "link": "{get_config("application.cookieConsent.linkText")}", "href": "{get_config("application.cookieConsent.linkUrl")}"}}'
    replace_in_file('frontend/dist/frontend/index.html', r'"content": { "message": ".*", "dismiss": ".*", "link": ".*", "href": ".*" }', content_property)

def is_url(url):
    return url.startswith('http')

def extract_filename(url):
    file = os.path.basename(url)
    if '?' in file:
        file = file.split('?')[0]
    return file

def download_to_file(url, dest):
    try:
        response = requests.get(url)
        response.raise_for_status()
        with open(dest, 'wb') as f:
            f.write(response.content)
    except requests.RequestException as e:
        logger.warn(f"Failed to download {url} ({str(e)})")

def replace_in_file(file_path, pattern, replacement):
    with open(file_path, 'r') as file:
        content = file.read()
    content_new = re.sub(pattern, replacement, content)
    with open(file_path, 'w') as file:
        file.write(content_new)
